Use C3 (mostly) to compute IRO.

jamadden opened the pull request:
Fixes #21

The 'mostly' is because interfaces are used in cases that C3 forbids; when there's a conflict, we move forward by adding the simple legacy base to the list at the conflicting location and continuing.

I hoped the fix for #8 might shake out automatically, but it didn't.

jensens commented:
Well, another recursive function. Bells are ringing. Did you test how this effects performance?

jamadden commented:
Yes. There is a slight overhead of around 10% or so. Elapsed time for loading ~440 `interface.py` modules from Plone and zopefoundation projects (a total of ~1900 interfaces) increases from around 5s to around 6s (there's a lot of variability in the test so I can't be more precise than that; some runs are better, some are worse). That doesn't seem like a problem to me (based on this one test), but if it winds up being so for whatever reason it's designed to be easy to implement in C (the CPython MRO C functions are nearly self contained and could be used with slight modifications).

Looking at those interfaces and declarations, there are, as expected, some IROs that change. (Of the 5600 ROs calculated I think that about 600 had differences, about 10%). They generally appear to make substantially more sense. For example, distant base classes tend to be towards the back, as one would expect. In this example, before, a generic adapter registration for `object` would override one for `IFileWidget`; that seems wrong:
```patch
 <implementedBy plone.formwidget.namedfile.widget.NamedFileWidget>
 (<implementedBy plone.formwidget.namedfile.widget.NamedFileWidget>,
  <implementedBy __builtin__.ExplicitAcquirer>,
  <InterfaceClass Acquisition.interfaces.IAcquirer>,
- <implementedBy ExtensionClass.Base>,
- <implementedBy __builtin__.object>,
  <implementedBy z3c.form.browser.file.FileWidget>,
  <InterfaceClass z3c.form.interfaces.IFileWidget>,
  <InterfaceClass z3c.form.interfaces.ITextWidget>,
  <InterfaceClass z3c.form.interfaces.IWidget>,
  <InterfaceClass zope.location.interfaces.ILocation>,
- <InterfaceClass zope.interface.Interface>)
+ <InterfaceClass zope.interface.Interface>,
+ <implementedBy ExtensionClass.Base>,
+ <implementedBy __builtin__.object>)
```

Others are a result of coming closer to implementing C3; usually I think the specific *C*onstraint that changes things is "monotonicity" ("The MRO of a class must be an extension without reordering of the MRO of each of its superclasses." The current algorithm is essentially breadth-first traversal with first-one-wins, meaning superclass MROs get re-ordered *all the time*). Given `class IFolder(IContainer, IPossibleSite, IAttributeAnnotatable)` on `class Folder(container.Folder, SiteManagerContainer)` where `container.Folder` actually implements `IContentContainer`, we now have the following order:
```patch
 <implementedBy zope.site.folder.Folder>
 (<implementedBy zope.site.folder.Folder>,
  <InterfaceClass zope.site.interfaces.IFolder>,
- <InterfaceClass zope.annotation.interfaces.IAttributeAnnotatable>,
- <InterfaceClass zope.annotation.interfaces.IAnnotatable>,
  <implementedBy zope.container.folder.Folder>,
  <InterfaceClass zope.container.interfaces.IContentContainer>,
  <implementedBy zope.container.btree.BTreeContainer>,
  <InterfaceClass zope.container.interfaces.IBTreeContainer>,
  <InterfaceClass zope.container.interfaces.IContainer>,
  <InterfaceClass zope.container.interfaces.IReadContainer>,
  <InterfaceClass zope.container.interfaces.ISimpleReadContainer>,
  <InterfaceClass zope.container.interfaces.IItemContainer>,
  <InterfaceClass zope.interface.common.mapping.IEnumerableMapping>,
  <InterfaceClass zope.interface.common.mapping.IReadMapping>,
  <InterfaceClass zope.interface.common.mapping.IItemMapping>,
  <ABCInterfaceClass zope.interface.common.collections.IContainer>,
  <ABCInterfaceClass zope.interface.common.collections.ISized>,
  <ABCInterfaceClass zope.interface.common.ABCInterface>,
  <InterfaceClass zope.container.interfaces.IWriteContainer>,
- <implementedBy persistent.Persistent>,
- <InterfaceClass persistent.interfaces.IPersistent>,
  <implementedBy zope.site.site.SiteManagerContainer>,
  <InterfaceClass zope.component.interfaces.IPossibleSite>,
+ <InterfaceClass zope.annotation.interfaces.IAttributeAnnotatable>,
+ <InterfaceClass zope.annotation.interfaces.IAnnotatable>,
  <implementedBy zope.container.contained.Contained>,
  <InterfaceClass zope.location.interfaces.IContained>,
  <InterfaceClass zope.location.interfaces.ILocation>,
+ <implementedBy persistent.Persistent>,
+ <InterfaceClass persistent.interfaces.IPersistent>,
  <InterfaceClass zope.interface.Interface>,
  <implementedBy __builtin__.object>)
```

I'm not fond of the first fallback to the legacy RO (`if base is None`) when a consistent C3 order couldn't be calculated (for one thing, it takes time). That path gets invoked *a lot*. Of 5,637 top-level calls to `ro` in the test, we need to go down the legacy path some 20,000 times — but that's only for 128 different objects. I wonder if there's something smarter that can be done.

jensens commented:
Thanks for the detailed explanation. Overall I think this is well thought and can be merged.

If it runs into a non-resolvable IRO, shouldn't this at least generate a warning? Or would this just generate to many?

jamadden commented:
> If it runs into a non-resolvable IRO, shouldn't this at least generate a warning? Or would this just generate to many?

I left  a TODO in the code about that because I just wasn't sure. My main concerns with issuing warnings are

1. it's not clear at what stacklevel to issue the warning (it really depends on why the RO is being calculated); and 
2. the warning can arise many times at places distant from the source of the actual problem (as the RO gets calculated for new subclasses or other changes); and
3. `warnings` caches whether to (re)issue a warning on the basis of stacklevel and message string, so if you include extra formatted text to make the error more clear, you blow up the cache; and
4. if we try to solve the first few problems, we wind up needing our own extra cache(s) (but an easy case that would reduce the noise would be to specially handle things when there's exactly one base; no need to recalculate the whole RO for that one base, we should just be able to copy it using `__iro__` and stick the subclass on the front: `if len(ob.__bases__) == 1: return [ob] + ob.__bases__[0].__iro__`. I should probably do that.)

Kind of what I'm imagining now (for some hypothetical future PR) is to have a way (environment variable) for a user to opt-in to having zope.interface track and/or warn about locations of inconsistent IROs. A user would know to turn this on because we would issue a simple warning just for the *first* inconsistent case we saw (so no worrying about stacklevels or extra messaging). The problem there, naturally, is that some of the common interfaces shipped in `zope.interface` itself exhibit inconsistent IROs right now, so until that's changed, everyone would always get a warning. (I should put that in a code comment).

jensens commented:
+1 for future implementation with opt-in

jamadden commented:
I think this is ready for review 😄 I went ahead and implemented the ideas for warnings and tracking of invalid resolution orders in this PR, and no warnings are issued by default in this PR. That's because both `zope.schema` and `zope.component` have classes that produce in inconsistent RO and will need to be fixed; `zope.testrunner` has those dependencies, so everyone that uses `zope.testrunner` would be seeing a warning they can do nothing about.

I looked at those two libraries, and it looks like updating them should be easy. It appears to be just some redundant declarations that can be removed.

jamadden commented:
Thanks @jensens!

> if it helps I can create a branch in Plones buildout.coredev using this branch to let a bigger test suite run with this.

If that's not too difficult, that sounds like it would be very helpful, just to make sure it doesn't break the world in a terribly hard to understand way.

I'm also thinking that I should add an environment variable that, when set, causes any differences between the legacy and new ROs to be discovered, like the other warning environment variables. That would help debug the situation if anyone runs into problems. (And, for completeness, probably an environment variable that forces the use of the legacy RO for the time being, so people can get the other advantages of what will be 5.0 if there are new ROs they can't adjust for immediately). I can do those in a followup PR if the Plone tests are clean enough; otherwise I can add them here.

jamadden commented:
I found https://github.com/plone/buildout.coredev/pull/647 which leads to the failing jenkins build at https://jenkins.plone.org/job/pull-request-5.2/1042/testReport/

I spot checked a very small number of the test failures. There seemed to be only a couple basic kinds (that weren't direct follow-ons from a previous proximate error or failed cleanup). [The first was an index error](https://jenkins.plone.org/job/pull-request-5.2/1042/testReport/junit/plone.restapi.tests.test_content_utils/TestAddContent/test_add_content_to_container_and_move_on_added_event/) `list index out of range`:  
```
File "/home/jenkins/.buildout/eggs/cp27m/Products.CMFCore-2.4.4-py2.7.egg/Products/CMFCore/CatalogTool.py", line 367, in _reindexObject
    self.catalog_object(object, uid, idxs, update_metadata)
  File "/home/jenkins/workspace/pull-request-5.2/src/Products.CMFPlone/Products/CMFPlone/CatalogTool.py", line 351, in catalog_object
    update_metadata, pghandler=pghandler)
  File "/home/jenkins/.buildout/eggs/cp27m/Products.ZCatalog-5.0.1-py2.7.egg/Products/ZCatalog/ZCatalog.py", line 499, in catalog_object
    update_metadata=update_metadata)
  File "/home/jenkins/.buildout/eggs/cp27m/Products.ZCatalog-5.0.1-py2.7.egg/Products/ZCatalog/Catalog.py", line 369, in catalogObject
    blah = x.index_object(index, object, threshold)
  File "/home/jenkins/.buildout/eggs/cp27m/Products.ExtendedPathIndex-3.4.2-py2.7.egg/Products/ExtendedPathIndex/ExtendedPathIndex.py", line 97, in index_object
    index = attrs is None and self.id or attrs[0]
```

The second [was an `AttributeError`]('TestAddContent' object has no attribute 'portal'):
```
  File "/home/jenkins/.buildout/eggs/cp27m/Zope-4.1.3-py2.7.egg/OFS/ObjectManager.py", line 384, in _setObject
    notify(ObjectAddedEvent(ob, self, id))
  File "/home/jenkins/.buildout/eggs/cp27m/zope.event-4.4-py2.7.egg/zope/event/__init__.py", line 32, in notify
    subscriber(event)
  File "/home/jenkins/.buildout/eggs/cp27m/zope.component-4.5-py2.7.egg/zope/component/event.py", line 27, in dispatch
    component_subscribers(event, None)
  File "/home/jenkins/.buildout/eggs/cp27m/zope.component-4.5-py2.7.egg/zope/component/_api.py", line 124, in subscribers
    return sitemanager.subscribers(objects, interface)
  File "/home/jenkins/workspace/pull-request-5.2/src/zope.interface/src/zope/interface/registry.py", line 448, in subscribers
    return self.adapters.subscribers(objects, provided)
  File "/home/jenkins/workspace/pull-request-5.2/src/zope.interface/src/zope/interface/adapter.py", line 619, in subscribers
    subscription(*objects)
  File "/home/jenkins/workspace/pull-request-5.2/src/plone.restapi/src/plone/restapi/tests/test_content_utils.py", line 138, in move_object
    self.portal.manage_pasteObjects(
```

I think those `AttributeErrors` are all caused by (exactly) that first test failure. That [test in plone.restapi fails to properly clean up after itself](https://github.com/plone/plone.restapi/blob/ebe264281c7d981efdcda1ff8d823dfe7df1d7ef/src/plone/restapi/tests/test_content_utils.py#L134-L149) and leaves a subscriber hanging around in the global site manager when it fails.

The first instance of the index test failure is actually caused by a `setUp` method:
```
Error in test test_DocumentEmptyLayoutBug (Products.CMFDynamicViewFTI.tests.test_fti.TestEmptyLayoutBug)
Traceback (most recent call last):
  File "/srv/python2.7/lib/python2.7/unittest/case.py", line 320, in run
    self.setUp()
  File "/home/jenkins/.buildout/eggs/cp27m/Products.CMFDynamicViewFTI-6.0.2-py2.7.egg/Products/CMFDynamicViewFTI/tests/test_fti.py", line 226, in setUp
    self.folder.invokeFactory('DynDocument', id='dyndocument')
…
  File "/home/jenkins/.buildout/eggs/cp27m/Products.ExtendedPathIndex-3.4.2-py2.7.egg/Products/ExtendedPathIndex/ExtendedPathIndex.py", line 97, in index_object
    index = attrs is None and self.id or attrs[0]
IndexError: list index out of range
```

Prior to that, there were only two or three test failures from the doctests for plone.schemaeditor:
```
File "/home/jenkins/.buildout/eggs/cp27m/plone.schemaeditor-2.1.0-py2.7.egg/plone/schemaeditor/tests/editing.rst", line 33, in editing.rst
Failed example:
    browser.open(portal_url + '/@@schemaeditor')
Exception raised:
    Traceback (most recent call last):
...
   File "/home/jenkins/workspace/pull-request-5.2/var/cache/fb50a9ac39cebc6c000b0c32fccbaff2.py", line 183, in render
        __macro = _static_139629253161872(getitem('context'), econtext, True, ('main_template', 'macros', 'master', ))
      File "/home/jenkins/.buildout/eggs/cp27m/Zope-4.1.3-py2.7.egg/Products/PageTemplates/expression.py", line 94, in __call__
        base = self.traverse(base, request, path_items)
      File "/home/jenkins/.buildout/eggs/cp27m/Zope-4.1.3-py2.7.egg/Products/PageTemplates/expression.py", line 62, in traverse
        base = traverser(name)
      File "/home/jenkins/.buildout/eggs/cp27m/Zope-4.1.3-py2.7.egg/OFS/Traversable.py", line 350, in unrestrictedTraverse
        raise e
       - __traceback_info__: ([], 'main_template')
    NotFound: main_template

     - Expression: "context/main_template/macros/master"
     - Filename:   ... rm-1.1.1-py2.7.egg/plone/z3cform/pagetemplates/layout.pt
     - Location:   (line 5: col 23)
     - Source:     ... etal:use-macro="context/main_template/macros/master">
                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     - Arguments:  repeat: {...} (0)
                   template: <ViewPageTemplateFile - at 0x7efdfba316d0>
                   views: <ViewMapper - at 0x7efdec783090>
                   modules: <_SecureModuleImporter - at 0x7efe03f4af50>
                   args: <tuple - at 0x7efe0ec9d050>
                   here: <ImplicitAcquisitionWrapper @@schemaeditor at 0x7efdf27ed0f0>
                   wrapped_repeat: <SafeMapping - at 0x7efdf70efaa0>
                   user: <ImplicitAcquisitionWrapper - at 0x7efdf27ed870>
                   nothing: <NoneType - at 0x563ffb69fae0>
                   root: <ImplicitAcquisitionWrapper  at 0x7efdee42c410>
                   container: <ImplicitAcquisitionWrapper @@schemaeditor at 0x7efdf27ed0f0>
                   request: <WSGIRequest - at 0x7efdec783110>
                   traverse_subpath: <list - at 0x7efdf4fc59b0>
                   default: <object - at 0x7efe09e9cc80>
                   context: <ImplicitAcquisitionWrapper @@schemaeditor at 0x7efdf27ed0f0>
                   view: <SchemaListingPage edit at 0x7efdec783cd0>
                   translate: <function translate at 0x7efdf0688550>
                   macroname: master
                   options: {...} (0)
                   loop: {...} (0)
                   target_language: <NoneType - at 0x563ffb69fae0>
```

If `context` is an `ImplicitAcquisitionWrapper` for a view, that's…probably unrelated? I don't really know enough about Plone to be able to interpret these results. What are your thoughts @jensens ?

jamadden commented:
I was able to clone `plone/buildout.coredev` and checkout the `test_zope_c3_iro` branch and dig into the most common issue, `IndexError: list index out of range`.

The trouble comes down to this `ExtendedPathIndex` having its `indexed_attrs` set to `[]` by `Products.GenericSetup.PluginIndexes.exportimport.PluggableIndexNodeAdapter`. That class adapts `IPluggableIndex`. There is a more specific adapter in that same module, `PathIndexNodeAdapter` registered for `IPathIndex`, that I would expect to get used.

Using `ZOPE_INTERFACE_LOG_CHANGED_IRO`, we see that `ExtendedPathIndex` did have a changed IRO, resulting in the less specific adapter being picked:
```
WARNING:zope.interface.ro:Object <implementedBy Products.ExtendedPathIndex.ExtendedPathIndex.ExtendedPathIndex> has different legacy and C3 MROs:
  Legacy RO (44)                                                                    C3 RO (44)
  ==================================================================================================================================================================
    <implementedBy Products.ExtendedPathIndex.ExtendedPathIndex.ExtendedPathIndex>    <implementedBy Products.ExtendedPathIndex.ExtendedPathIndex.ExtendedPathIndex>
    <InterfaceClass Products.PluginIndexes.interfaces.ILimitedResultIndex>            <InterfaceClass Products.PluginIndexes.interfaces.ILimitedResultIndex>
                                                                                    + <InterfaceClass Products.PluginIndexes.interfaces.IPluggableIndex>
                                                                                    + <InterfaceClass zope.interface.Interface>
                                                                                    + <InterfaceClass Products.PluginIndexes.interfaces.IQueryIndex>
    <implementedBy Products.PluginIndexes.PathIndex.PathIndex.PathIndex>              <implementedBy Products.PluginIndexes.PathIndex.PathIndex.PathIndex>
    <InterfaceClass Products.PluginIndexes.interfaces.IPathIndex>                     <InterfaceClass Products.PluginIndexes.interfaces.IPathIndex>
  - <InterfaceClass Products.PluginIndexes.interfaces.IQueryIndex>
    <InterfaceClass Products.PluginIndexes.interfaces.IUniqueValueIndex>              <InterfaceClass Products.PluginIndexes.interfaces.IUniqueValueIndex>
    <InterfaceClass Products.PluginIndexes.interfaces.ISortIndex>                     <InterfaceClass Products.PluginIndexes.interfaces.ISortIndex>
  - <InterfaceClass Products.PluginIndexes.interfaces.IPluggableIndex>
    <implementedBy OFS.SimpleItem.SimpleItem>                                         <implementedBy OFS.SimpleItem.SimpleItem>
    <InterfaceClass OFS.interfaces.ISimpleItem>                                       <InterfaceClass OFS.interfaces.ISimpleItem>
    <implementedBy OFS.SimpleItem.Item>                                               <implementedBy OFS.SimpleItem.Item>
    <InterfaceClass OFS.interfaces.IItem>                                             <InterfaceClass OFS.interfaces.IItem>
    <InterfaceClass OFS.interfaces.IZopeObject>                                       <InterfaceClass OFS.interfaces.IZopeObject>
    <InterfaceClass OFS.interfaces.IManageable>                                       <InterfaceClass OFS.interfaces.IManageable>
    <implementedBy OFS.SimpleItem.PathReprProvider>                                   <implementedBy OFS.SimpleItem.PathReprProvider>
                                                                                    + <InterfaceClass OFS.interfaces.ICopySource>
                                                                                    + <InterfaceClass OFS.interfaces.ITraversable>
                                                                                    + <InterfaceClass AccessControl.interfaces.IOwned>
                                                                                    + <implementedBy Persistence.Persistent>
                                                                                    + <implementedBy persistent.Persistent>
                                                                                    + <InterfaceClass persistent.interfaces.IPersistent>
                                                                                    + <implementedBy __builtin__.Acquirer>
                                                                                    + <InterfaceClass Acquisition.interfaces.IAcquirer>
                                                                                    + <implementedBy OFS.role.RoleManager>
                                                                                    + <implementedBy AccessControl.rolemanager.RoleManager>
                                                                                    + <InterfaceClass AccessControl.interfaces.IRoleManager>
                                                                                    + <InterfaceClass AccessControl.interfaces.IPermissionMappingSupport>
    <implementedBy App.Management.Navigation>                                         <implementedBy App.Management.Navigation>
    <InterfaceClass App.interfaces.INavigation>                                       <InterfaceClass App.interfaces.INavigation>
    <implementedBy webdav.Resource.Resource>                                          <implementedBy webdav.Resource.Resource>
    <InterfaceClass webdav.interfaces.IDAVResource>                                   <InterfaceClass webdav.interfaces.IDAVResource>
    <implementedBy OFS.Lockable.LockableItem>                                         <implementedBy OFS.Lockable.LockableItem>
    <InterfaceClass OFS.interfaces.IWriteLock>                                        <InterfaceClass OFS.interfaces.IWriteLock>
    <implementedBy OFS.EtagSupport.EtagSupport>                                       <implementedBy OFS.EtagSupport.EtagSupport>
    <InterfaceClass OFS.EtagSupport.EtagBaseInterface>                                <InterfaceClass OFS.EtagSupport.EtagBaseInterface>
    <implementedBy OFS.CopySupport.CopySource>                                        <implementedBy OFS.CopySupport.CopySource>
  - <InterfaceClass OFS.interfaces.ICopySource>
    <implementedBy App.Management.Tabs>                                               <implementedBy App.Management.Tabs>
    <implementedBy OFS.Traversable.Traversable>                                       <implementedBy OFS.Traversable.Traversable>
  - <InterfaceClass OFS.interfaces.ITraversable>
    <implementedBy OFS.owner.Owned>                                                   <implementedBy OFS.owner.Owned>
    <implementedBy AccessControl.owner.Owned>                                         <implementedBy AccessControl.owner.Owned>
  - <InterfaceClass AccessControl.interfaces.IOwned>
  - <implementedBy Persistence.Persistent>
  - <implementedBy persistent.Persistent>
  - <InterfaceClass persistent.interfaces.IPersistent>
  - <implementedBy __builtin__.Acquirer>
  - <InterfaceClass Acquisition.interfaces.IAcquirer>
  - <implementedBy OFS.role.RoleManager>
  - <implementedBy AccessControl.rolemanager.RoleManager>
  - <InterfaceClass AccessControl.interfaces.IRoleManager>
    <implementedBy ExtensionClass.Base>                                               <implementedBy ExtensionClass.Base>
    <implementedBy __builtin__.object>                                                <implementedBy __builtin__.object>
    <implementedBy AccessControl.PermissionMapping.RoleManager>                       <implementedBy AccessControl.PermissionMapping.RoleManager>
  - <InterfaceClass AccessControl.interfaces.IPermissionMappingSupport>
  - <InterfaceClass zope.interface.Interface>
```

Github inexplicably constrains comments to be ridiculously narrow, so to see that without horizontal scrolling, the IRO *used* to start:
``` 
 <implementedBy ...ExtendedPathIndex.ExtendedPathIndex
 <InterfaceClass ...PluginIndexes.interfaces.ILimitedResultIndex>        
 <implementedBy ...PluginIndexes.PathIndex.PathIndex.PathIndex>         
 <InterfaceClass ...PluginIndexes.interfaces.IPathIndex>                
- <InterfaceClass Products.PluginIndexes.interfaces.IQueryIndex>               
 <InterfaceClass Products.PluginIndexes.interfaces.IUniqueValueIndex>         
 <InterfaceClass Products.PluginIndexes.interfaces.ISortIndex>                
- <InterfaceClass Products.PluginIndexes.interfaces.IPluggableIndex>           
 <implementedBy OFS.SimpleItem.SimpleItem>                                    
 <InterfaceClass OFS.interfaces.ISimpleItem>                                  
```

But it is now:
```
   <implementedBy ...ExtendedPathIndex.ExtendedPathIndex
   <InterfaceClass ..PluginIndexes.interfaces.ILimitedResultIndex>       
 + <InterfaceClass ..PluginIndexes.interfaces.IPluggableIndex>           
 + <InterfaceClass zope.interface.Interface>                                    
 + <InterfaceClass Products.PluginIndexes.interfaces.IQueryIndex>               
   <implementedBy Products.PluginIndexes.PathIndex.PathIndex.PathIndex>         
   <InterfaceClass Products.PluginIndexes.interfaces.IPathIndex>                
   <InterfaceClass Products.PluginIndexes.interfaces.IUniqueValueIndex>         
   <InterfaceClass Products.PluginIndexes.interfaces.ISortIndex>                
   <implementedBy OFS.SimpleItem.SimpleItem>                                    
```

Why the change?

First, the interface definition tree:
```
Interface
\
 |- IPluggableIndex
 | \
 |  |- ILimitedResultIndex
 |  |- IQueryIndex
 |- IPathIndex
```

Next, the class definitions:
```python
@implementer(IPathIndex, IQueryIndex, IUniqueValueIndex, ISortIndex)
class PathIndex(…): pass

@implementer(ILimitedResultIndex, IQueryIndex)
class ExtendedPathIndex(PathIndex): …
```

The new order is closer to what the author actually declared.

The way `@implementer` works is to put those interfaces specifically given at the *beginning* of the `__bases__`. Thus, the `__bases__` for `<implementedBy ExtendedPathIndex>` are `(ILimitedResultIndex, IQueryIndex, <implementedBy PathIndex>)`, and we have to attempt to come up with some meaningful resolution  order from that.

Those three, respectively, have their own `__sro__`:
```python
[[#ILimitedResultIndex 
  <InterfaceClass Products.PluginIndexes.interfaces.ILimitedResultIndex>,
  <InterfaceClass Products.PluginIndexes.interfaces.IPluggableIndex>,
  <InterfaceClass zope.interface.Interface>],
 [#IQueryIndex
  <InterfaceClass Products.PluginIndexes.interfaces.IQueryIndex>,
  <InterfaceClass Products.PluginIndexes.interfaces.IPluggableIndex>,
  <InterfaceClass zope.interface.Interface>],
 [#Pathindex
  <implementedBy Products.PluginIndexes.PathIndex.PathIndex.PathIndex>,
  <InterfaceClass Products.PluginIndexes.interfaces.IPathIndex>,
  <InterfaceClass Products.PluginIndexes.interfaces.IQueryIndex>,
  <InterfaceClass Products.PluginIndexes.interfaces.IUniqueValueIndex>,
  <InterfaceClass Products.PluginIndexes.interfaces.ISortIndex>,
  <InterfaceClass Products.PluginIndexes.interfaces.IPluggableIndex>,
  <InterfaceClass zope.interface.Interface>,
  # ...
  <implementedBy __builtin__.object>,
  <implementedBy AccessControl.PermissionMapping.RoleManager>],]
```

First, we choose `ILimitedResultIndex`. It's the first declared base, and nobody else declares it anywhere else, so it becomes the first item in the RO. 

Long story short, the next two objects are completely inconsistent (they're not in the same order in any of the lists) so we're forced into the fallback case, and even the fallback-to-the-fallback and just wind up choosing them in that order. Again, this matches what the author declared, **but maybe our heuristics can be better.**

The main inconsistency, here, BTW, that causes the problem, is that `ExtendedPathIndex` tries to put `IQueryIndex` *ahead* of `PathIndex` while also inheriting from `PathIndex`, which itself 
brings in `IQueryIndex`. That's like these class definitions:
```pycon
In [1]: class Interface(object): pass
In [2]: class Pluggable(Interface): pass
In [3]: class Limited(Pluggable): pass
In [4]: class Query(Pluggable): pass
In [5]: class PathIndex(Interface): pass
In [6]: class Path(PathIndex, Query): pass
In [7]: class Extended(Limited, Query, Path): pass
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-8-95f8ff652f7d> in <module>()
----> 1 class Extended(Limited, Query, Path): pass

TypeError: Error when calling the metaclass bases
    Cannot create a consistent method resolution
order (MRO) for bases Pluggable, Query, Path
```

Dropping the redeclaration of `IQueryIndex` solves the inconsistency and lets the tests work, just as it lets the class be created. (Setting `ZOPE_INTERFACE_USE_LEGACY_IRO` also lets the tests pass, albeit verbosely.)
```pycon
In [8]: class Extended(Limited, Path): pass

```

jensens commented:
> Dropping the redeclaration of IQueryIndex solves the inconsistency and lets the tests work, just as it lets the class be created. (Setting ZOPE_INTERFACE_USE_LEGACY_IRO also lets the tests pass, albeit verbosely.)

I created a branch of ExtendedPathIndex with the fix and started the tests again.

jensens commented:
This class of tests is fixed! `Test Result (139 failures / -1152)`.

jensens commented:
If the Python 2.7 test run goes green too https://jenkins.plone.org/job/pull-request-5.2/1046/ 
then from my perspective this one is ready for merge!

jensens commented:
tests are running again here https://github.com/plone/buildout.coredev/pull/644
if green I from my perspective would consider this PR ready for merge

jamadden commented:
I've been iterating on this locally. I haven't been able to find a heuristic that gets all of the plone tests to pass at the same level they do against zope.interface master. I think that's because some of the objects were winding up with resolution orders that are *really* borked up (things like `object` somewhere up in the middle and stuff like that). 

I think we're going to have to give up on the idea of "partial consistency" and instead of using a different heuristic to pick the next base in the event of an inconsistency, just give up and use the legacy RO. Fortunately, there are only a few objects that are messed up that way. Out of 1800 static interface specifications, there were only 99 that I detected as inconsistent with C3 (and only 79 that changed from legacy to C3). 

I'm running the plone suite again with the fallback-to-legacy path back in place. When that's done, I should know exactly how many *dynamic* specifications changed (and how many had to fallback), and hopefully I'll know that no extra tests failed locally. I'll update this when I get that info.

jamadden commented:
I've pushed a change that simply falls back to the old order instead of trying to reach something closer to a C3 order when C3 fails. 

I've been running the `buildout.coredev` tests, using the `test_zope_c3_iro` branch.

First, I switched it back to zope.interface 4.6.0. It ran 10166 tests with 4 failures. Then, I switched it to zope.interface master, and 20 tests failed. With this latest commit on the branch in this PR, I got only 6 failures.
```
Total: 10166 tests, 6 failures, 1 errors, 63 skipped in 25 minutes 24.670 seconds.
```

I tracked the number of calls to `ro.ro` and collected stats during the test run. There were 148,868 top-level RO requests. 138,461 produced an order equal to the older order, while 10,407 produced a different order. 12,934 inconsistencies were detected at the top level and used the old algorithm.

jamadden commented:
Argh. I forgot that the `buildout.coredev` test suite is sensitive to `$PYTHONHASHSEED`. I default to having that set to `random`, but the test suite is sensitive to things that change when hashes change under Py2.7 (dict/set iteration order).

Before, I used several different terminal sessions to run tests and didn't have the variable undefined consistently in all of them. I re-ran the tests in a single session with `$PYTHONHASHSEED` undefined. With zope.interface 4.6, I get 4 failures, 1 error — all due to not having some binaries installed that the tests look for. Switching to master, I get one additional failure of `plone.app.relationfield-2.0.1-py2.7.egg/plone/app/relationfield/tests/../schemaeditor.txt`. 

Switching to this branch, there's one new failure in `plone.app.dexterity-2.6.4-py2.7.egg/plone/app/dexterity/tests/editing.rst`, "    ComponentLookupError: ((<zope.schema._bootstrapfields.Field object at 0x107aafe90 IMinMax.min>, <WSGIRequest, URL=None>), <InterfaceClass z3c.form.interfaces.IFieldWidget>, u'')" at "/z3c/form/field.py", line 267, in update". The difference between master and this branch seems to be that master never gets the `IMinMax.min` field; instead, it gets `plone.schemaeditor.schema.IDate.min`. I don't know why yet. (One difference: in this branch, the form's `_schema` is `plone.app.z3cform.interfaces.IDateField` while in master it's `plone.schemaeditor.schema.IDate` for the same URL, "http://nohost/plone/dexterity-types/plonista/zope_schema__field_date/@@edit".)

jamadden commented:
Found it. The difference is in plone.schemaeditor.fields:
```python
@interface.implementer(interfaces.IFieldEditFormSchema)
@component.adapter(schema_ifaces.IField)
def getFirstFieldSchema(field):
    return [s for s in field.__provides__.__iro__ if
            s.isOrExtends(schema_ifaces.IField)][0]
```

In master, the field's IRO is:
```
[<InterfaceClass plone.schemaeditor.browser.field.edit.IFieldProxy>, 
 <InterfaceClass plone.schemaeditor.schema.IDate>, # Extends z.s.IDate
 <InterfaceClass plone.app.z3cform.interfaces.IDateField>, # Extends z.s.IDate
 <InterfaceClass zope.schema.interfaces.IDate>, 
 <InterfaceClass zope.schema.interfaces.IMinMax>,
  <InterfaceClass zope.schema.interfaces.IOrderable>, 
 <InterfaceClass zope.schema.interfaces.IField>, 
 <InterfaceClass zope.schema._bootstrapinterfaces.IValidatable>, 
 <InterfaceClass zope.interface.interfaces.IAttribute>, 
 <InterfaceClass zope.interface.interfaces.IElement>, 
 <InterfaceClass zope.interface.Interface>]
```

While here it is 
```
[<InterfaceClass plone.schemaeditor.browser.field.edit.IFieldProxy>, 
 <InterfaceClass plone.app.z3cform.interfaces.IDateField>, 
 <InterfaceClass plone.schemaeditor.schema.IDate>, 
 <InterfaceClass zope.schema.interfaces.IDate>, 
 <InterfaceClass zope.schema.interfaces.IMinMax>, 
 <InterfaceClass zope.schema.interfaces.IOrderable>,
  <InterfaceClass zope.schema.interfaces.IField>, 
 <InterfaceClass zope.schema._bootstrapinterfaces.IValidatable>, 
 <InterfaceClass zope.interface.interfaces.IAttribute>, 
 <InterfaceClass zope.interface.interfaces.IElement>, 
 <InterfaceClass zope.interface.Interface>]
```

The problem is that both `plone.app.z3cform.interfaces.IDateField` and `plone.schemaeditor.schema.IDate` both directly extend `zope.schema.interfaces.IDate`, and then `zope.schema._field.Date` is declared to implement `plone.schemaeditor.schema.IDate` in ZCML.  `plone.schemaeditor.schema` is aware of the ambiguity and tries to workaround it:
```python
try:
    from plone.app.z3cform.widgets import IDatetimeField as IDatetime
    from plone.app.z3cform.widgets import IDateField as IDate
    IDatetime, IDate  # pyflakes
except ImportError:
    try:
        from collective.z3cform.datetimewidget.interfaces import (
            IDatetimeField as IDatetime)
        from collective.z3cform.datetimewidget.interfaces import (
            IDateField as IDate)
        IDatetime, IDate  # pyflakes
    except ImportError:
        IDatetime = interfaces.IDatetime
        IDate = interfaces.IDate
```

Unfortunately that first import fails, and we wind up with the flat structure. 

If I change it to the correct import (`from plone.app.z3cform.interfaces`), the test passes!

jamadden commented:
@jensens Do you feel this can be merged?

jensens commented:
If [fixed plone.schemaeditor](https://github.com/plone/plone.schemaeditor/pull/74) after your findings. 

From my POV this can be merged, all possible upcoming problems are fixable as far as I can see. That said, it is good to be able to detect those problems now! Thanks for the great work!

jamadden commented:
Thanks!
